client/consul_template.go

```diff
@@ -0,0 +1,446 @@
+package client
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	ctconf "github.com/hashicorp/consul-template/config"
+	"github.com/hashicorp/consul-template/manager"
+	"github.com/hashicorp/consul-template/signals"
+	"github.com/hashicorp/consul-template/watch"
+	"github.com/hashicorp/nomad/client/config"
+	"github.com/hashicorp/nomad/client/driver/env"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+var (
+	// testRetryRate is used to speed up tests by setting consul-templates retry
+	// rate to something low
+	testRetryRate time.Duration = 0
+)
+
+// TaskHooks is an interface which provides hooks into the tasks life-cycle
+type TaskHooks interface {
+	// Restart is used to restart the task
+	Restart()
+
+	// Signal is used to signal the task
+	Signal(os.Signal)
+
+	// UnblockStart is used to unblock the starting of the task. This should be
+	// called after prestart work is completed
+	UnblockStart()
+
+	// Kill is used to kill the task because of the passed error.
+	Kill(error)
+}
+
+// TaskTemplateManager is used to run a set of templates for a given task
+type TaskTemplateManager struct {
+	// templates is the set of templates we are managing
+	templates []*structs.Template
+
+	// lookup allows looking up the set of Nomad templates by their consul-template ID
+	lookup map[string][]*structs.Template
+
+	// allRendered marks whether all the templates have been rendered
+	allRendered bool
+
+	// hooks is used to signal/restart the task as templates are rendered
+	hook TaskHooks
+
+	// runner is the consul-template runner
+	runner *manager.Runner
+
+	// signals is a lookup map from the string representation of a signal to its
+	// actual signal
+	signals map[string]os.Signal
+
+	// shutdownCh is used to signal and started goroutine to shutdown
+	shutdownCh chan struct{}
+
+	// shutdown marks whether the manager has been shutdown
+	shutdown     bool
+	shutdownLock sync.Mutex
+}
+
+func NewTaskTemplateManager(hook TaskHooks, tmpls []*structs.Template,
+	allRendered bool, config *config.Config, vaultToken, taskDir string,
+	taskEnv *env.TaskEnvironment) (*TaskTemplateManager, error) {
+
+	// Check pre-conditions
+	if hook == nil {
```

Comment:
This will miss the case of someone creating a var like `var hook *MyTaskHook`, never initializing it, and passing it in. It will be a non-nil interface with a nil value.

I don't think it's worth trying to guard against that case, but I thought it was worth pointing out these kinds of guards are imperfect.